src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -1346,8 +1440,14 @@ CleanUpSessions(DistributedExecution *execution)
 			 * We cannot get MULTI_CONNECTION_LOST via the ConnectionStateMachine,
 			 * but we might get it via the connection API and find us here before
 			 * changing any states in the ConnectionStateMachine.
+			 *
+			 * If execution is outside a transaction, then the connection will be shutdown later so
+			 * we can skip it here.
 			 */
-			CloseConnection(connection);
+			if (!execution->isOutsideTransaction)
+			{
```

Comment:
Could we just let it close here? (We should perhaps introduce a MULTI_CONNECTION_CLOSED)

Reply:
What value would introducing `MULTI_CONNECTION_CLOSED`  have? We are freeing the connection in https://github.com/citusdata/citus/blob/master/src/backend/distributed/connection/connection_management.c#L458 while closing the connection. So even if we set that value it will not be `MULTI_CONNECTION_CLOSED`.

Reply:
Let me rephrase: What happens if we just call CloseConnection as before instead of doing this check?

Reply:
As far as I see we do not remove a pointer from the `sessionList`. So when we close a connection it gets freed but the pointer is still there pointing to something random. In that case when we iterate the `sessionList` again later we still have the connections in the session list. 

I guess the problem happens when the connection status is `connecting` so in that case after we are done with the execution, adaptive executor tries to close the connections that are either `connecting`, `lost`, or `failed`. I am not sure why we are not removing the session from `sessionList` when its connection is closed. Is there a reason for that? Maybe I am missing where it is being removed. But if it was we wouldnt get the error I described.